grant_disbursement/src/main.leo

```diff
@@ -0,0 +1,255 @@
+import credits.aleo;
+import multi_token_support_program.aleo;
+import pondo_core_protocol.aleo;
+
+// The 'grant_disbursement' program.
+program grant_disbursement.aleo {
+  struct Grant {
+    credits_amount: u64, // The amount of paleo to be distributed
+    paleo_amount: u64, // The amount of paleo to be distributed
+    recipient_rewards_key: address, // The address that can withdraw the rewards
+    recipient_principal_key: address, // The address that can withdraw the principal
+    start_block: u32, // The block at which the grant was created
+    cliff_block: u32, // The block at which the recipient can start withdrawing principal
+    fully_vested_block: u32 // The block at which the recipient can withdraw the full principal
+  }
+
+  mapping grants: u64 => Grant;
+
+  async transition create_grant(
+    public id: u64,
+    public credits_amount: u64,
+    public paleo_amount: u64,
+    public recipient_rewards_key: address,
+    public recipient_principal_key: address,
+    public cliff_block: u32,
+    public fully_vested_block: u32
+  ) -> Future {
+    // Deposit the credits into the pondo core protocol
+    let f0: Future = pondo_core_protocol.aleo/deposit_public_as_signer(credits_amount, paleo_amount, self.address);
+    // Transfer the pAleo to the grant disbursement program
+    let f1: Future = multi_token_support_program.aleo/transfer_public_as_signer(PALEO_TOKEN_ID, self.address, paleo_amount as u128);
+
+    assert(fully_vested_block > cliff_block);
+
+    return finalize_create_grant(
+      f0,
+      f1,
+      id,
+      credits_amount,
+      paleo_amount,
+      recipient_rewards_key,
+      recipient_principal_key,
+      cliff_block,
+      fully_vested_block
+      );
+  }
+
+  async function finalize_create_grant(
+    f0: Future,
+    f1: Future,
+    id: u64,
+    credits_amount: u64,
+    paleo_amount: u64,
+    recipient_rewards_key: address,
+    recipient_principal_key: address,
+    cliff_block: u32,
+    fully_vested_block: u32
+  ) {
+    f0.await();
+    f1.await();
+
+    let grant: Grant = Grant {
+      credits_amount: credits_amount,
+      paleo_amount: paleo_amount,
+      recipient_rewards_key: recipient_rewards_key,
+      recipient_principal_key: recipient_principal_key,
+      start_block: block.height,
```

Comment:
Not sure if this is the way we want to go, but allowing the block height to be passed in rather than taking `block.height` would allow grants to be set up to start vesting at a given height.